File: src/neuron.c
/**
 * File : 	neuron.c
 * 	
 * Description : Contain the neuron structure and related functions
 *		
 * Authors :
 *		Cuartero Jean-Louis
 * 		Delpech Marc	
 * 		Tissier Romain
 */
#include "neuron.h"
#include <stdlib.h>

/**
 * Constructor creating an empty neuron object
 */
Neuron *neuron_create()
{
	Neuron *neuron = malloc(sizeof(Neuron));
	neuron->value = 0;
	neuron->gradient = 0;
	return neuron;
}

/**
 * Destructor for a neuron object
 *	Param : 
 *		neuron : neuron object to destroy
 */
void neuron_destroy(Neuron * neuron)
{
	// TODO free cross referenced links
	free(neuron);
}

File: src/Neuron.java
import java.io.Serializable;
import java.util.ArrayList;

/**
 * Class representing a neuron within the network
 * 
 * @authors Cuartero Jean-Louis, Delpech Marc, Tissier Romain
 * 
 */
public class Neuron implements Serializable {
	private ArrayList<Link> inputs;
	private ArrayList<Link> outputs;
	private Double gradient;
	private Double value;

	/**
	 * Constructor initializing a neuron with empty inputs and outputs and 0
	 * gradient and value
	 */
	public Neuron() {
		inputs = new ArrayList<Link>();
		outputs = new ArrayList<Link>();
		gradient = 0.0;
		value = 0.0;
	}

	/**
	 * Getter returning the input list of the neuron
	 * 
	 * @return the input list of the neuron
	 */
	public ArrayList<Link> getInputs() {
		return inputs;
	}

	/**
	 * Setter updating the input list of the neuron
	 * 
	 * @param inputs
	 *            new input list
	 */
	public void setInputs(ArrayList<Link> inputs) {
		this.inputs = inputs;
	}

	/**
	 * Getter returning the output list of the neuron
	 * 
	 * @return the output list of the neuron
	 */
	public ArrayList<Link> getOutputs() {
		return outputs;
	}

	/**
	 * Setter updating the output list of the neuron
	 * 
	 * @param outputs
	 *            new output list
	 */
	public void setOutputs(ArrayList<Link> outputs) {
		this.outputs = outputs;
	}

	/**
	 * Getter returning the gradient value of the neuron
	 * 
	 * @return the gradient value of the neuron
	 */
	public Double getGradient() {
		return gradient;
	}

	/**
	 * Updating the gradient value
	 * 
	 * @param gradient
	 *            new value
	 */
	public void setGradient(Double gradient) {
		this.gradient = gradient;
	}

	/**
	 * Getter returning the value (issued from activation function) of the
	 * neuron
	 * 
	 * @return the value of the neuron
	 */
	public Double getValue() {
		return value;
	}

	/**
	 * Update the value of the neuron
	 * 
	 * @param value
	 *            new value
	 */
	public void setValue(Double value) {
		this.value = value;
	}
}
